pytorch_lightning/accelerators/gpu.py

```diff
@@ -53,6 +57,78 @@ def set_nvidia_flags(local_rank: int) -> None:
         devices = os.getenv("CUDA_VISIBLE_DEVICES", all_gpu_ids)
         _log.info(f"LOCAL_RANK: {local_rank} - CUDA_VISIBLE_DEVICES: [{devices}]")
 
+    def get_device_stats(self, device: Union[str, torch.device]) -> Dict[str, Any]:
+        """Gets stats for the given GPU device.
+
+        Returns:
+            A dictionary mapping the metrics to their values.
+
+        Raises:
+            FileNotFoundError:
+                If nvidia-smi installation not found
+        """
+        if _TORCH_GREATER_EQUAL_1_8:
+            return torch.cuda.memory_stats(device)
+        else:
+            return _get_nvidia_gpu_stats(device)
+
     def teardown(self) -> None:
         super().teardown()
         self._move_optimizer_state(torch.device("cpu"))
+
+
+def _get_nvidia_gpu_stats(device: torch.device) -> Dict[str, float]:
+    """Get GPU stats including memory, fan speed, and temperature from nvidia-smi.
+
+    Returns:
+        A dictionary mapping the metrics to their values.
+
+    Raises:
+        FileNotFoundError:
+            If nvidia-smi installation not found
+    """
+    gpu_stat_metrics = [
+        ("utilization.gpu", "%"),
+        ("memory.used", "MB"),
+        ("memory.free", "MB"),
+        ("utilization.memory", "%"),
+        ("fan.speed", "%"),
+        ("temperature.gpu", "°C"),
+        ("temperature.memory", "°C"),
+    ]
+    gpu_stat_keys = [k for k, _ in gpu_stat_metrics]
+    gpu_query = ",".join(gpu_stat_keys)
+
+    gpu_id = _get_gpu_id(device.index)
+    nvidia_smi_path = shutil.which("nvidia-smi")
+    if nvidia_smi_path is None:
+        raise FileNotFoundError("nvidia-smi: command not found")
```

Comment:
just curious what is the case of having GPU support and not SMI? is it with Conda and `cudatoolkit`?
eventually, do we need some min SMI version?

Reply:
Hmm not sure about the answer to this. maybe @rohitgr7 knows since he worked on this in the past?

Reply:
@Borda it is possible for nvidia-smi to not be available in the image. this is the case for one of our clusters. sharing some motivations for why it was removed:

- it has undesired side-effects that we had to patch (like loading the module, creating all the devices because by default it's setuid)
- it causes slowness and potentially deadlocks as we have more than one process doing NVML calls
- it can be confusing, like the version of CUDA shown in its output is the latest supported version as per the nvidia-smi binary, NOT what is actually installed / available on the host
- it requires being in sync'ed between the host and the image, which is a really hard thing to maintain. nvidia-smi requires a lot of NVIDIA packages to be installed, which would be really hard to expose through the container.

Reply:
back then I was just testing the `GPUStatsMonitor` on Kaggle/Colab GPU (can't remember exactly) and faced this issue sometimes in their sessions. so added this check as well.